src/rabbit_core_metrics.erl

```diff
@@ -130,35 +146,59 @@ channel_stats(reductions, Id, Value) ->
     ok.
 
 channel_stats(exchange_stats, publish, Id, Value) ->
-    ets:update_counter(channel_exchange_metrics, Id, {2, Value}, {Id, 0, 0, 0}),
+    ets_update_counter(channel_exchange_metrics, Id, {2, Value}, {Id, 0, 0, 0}),
     ok;
 channel_stats(exchange_stats, confirm, Id, Value) ->
-    ets:update_counter(channel_exchange_metrics, Id, {3, Value}, {Id, 0, 0, 0}),
+    ets_update_counter(channel_exchange_metrics, Id, {3, Value}, {Id, 0, 0, 0}),
     ok;
 channel_stats(exchange_stats, return_unroutable, Id, Value) ->
-    ets:update_counter(channel_exchange_metrics, Id, {4, Value}, {Id, 0, 0, 0}),
+    ets_update_counter(channel_exchange_metrics, Id, {4, Value}, {Id, 0, 0, 0}),
     ok;
 channel_stats(queue_exchange_stats, publish, Id, Value) ->
-    ets:update_counter(channel_queue_exchange_metrics, Id, Value, {Id, 0}),
+    ets_update_counter(channel_queue_exchange_metrics, Id, Value, {Id, 0}),
     ok;
 channel_stats(queue_stats, get, Id, Value) ->
-    ets:update_counter(channel_queue_metrics, Id, {2, Value}, {Id, 0, 0, 0, 0, 0, 0}),
+    ets_update_counter(channel_queue_metrics, Id, {2, Value}, {Id, 0, 0, 0, 0, 0, 0}),
     ok;
 channel_stats(queue_stats, get_no_ack, Id, Value) ->
-    ets:update_counter(channel_queue_metrics, Id, {3, Value}, {Id, 0, 0, 0, 0, 0, 0}),
+    ets_update_counter(channel_queue_metrics, Id, {3, Value}, {Id, 0, 0, 0, 0, 0, 0}),
     ok;
 channel_stats(queue_stats, deliver, Id, Value) ->
-    ets:update_counter(channel_queue_metrics, Id, {4, Value}, {Id, 0, 0, 0, 0, 0, 0}),
+    ets_update_counter(channel_queue_metrics, Id, {4, Value}, {Id, 0, 0, 0, 0, 0, 0}),
     ok;
 channel_stats(queue_stats, deliver_no_ack, Id, Value) ->
-    ets:update_counter(channel_queue_metrics, Id, {5, Value}, {Id, 0, 0, 0, 0, 0, 0}),
+    ets_update_counter(channel_queue_metrics, Id, {5, Value}, {Id, 0, 0, 0, 0, 0, 0}),
     ok;
 channel_stats(queue_stats, redeliver, Id, Value) ->
-    ets:update_counter(channel_queue_metrics, Id, {6, Value}, {Id, 0, 0, 0, 0, 0, 0}),
+    ets_update_counter(channel_queue_metrics, Id, {6, Value}, {Id, 0, 0, 0, 0, 0, 0}),
     ok;
 channel_stats(queue_stats, ack, Id, Value) ->
-    ets:update_counter(channel_queue_metrics, Id, {7, Value}, {Id, 0, 0, 0, 0, 0, 0}),
-    ok.
+    ets_update_counter(channel_queue_metrics, Id, {7, Value}, {Id, 0, 0, 0, 0, 0, 0}),
+    ok.
+
+%% ets:update_counter(Tab, Key, Incr, Default) appeared in Erlang 18.x.
+%% We need a wrapper for Erlang R16B03 and Erlang 17.x. This wrapper
+%% first creates the entry if it's missing, then updates it.
+
+ets_update_counter(Tab, Key, Incr, Default) ->
+    code_version:update(?MODULE),
+    ?MODULE:ets_update_counter(Tab, Key, Incr, Default).
+
+ets_update_counter_pre_18(Tab, Key, Incr, Default) ->
+    ets:insert_new(Tab, Default),
```

Comment:
Not strictly.`ets:update_counter/4` always inserts the value if they key is not there, thus `update_counter/4` always succeeds. Now, we are allowing `0`s (let's do nothing) which weren't possible before.